notebooks/nessie-iceberg-flink-demo-nba.ipynb

```diff
@@ -0,0 +1,719 @@
+    "To get started, we will first have to do a few setup steps that give us everything we need\n",
+    "to get started with Nessie. In case you're interested in the detailed setup steps for Flink, you can check out the [docs](https://projectnessie.org/tools/flink/)\n",
+    "\n",
+    "The Binder server has downloaded flink and some data for us as well as started a Nessie server in the background. All we have to do is start Flink\n",
+    "\n",
+    "The below cell starts a local Flink session with parameters needed to configure Nessie. Each config option is followed by a comment explaining its purpose.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "from pyflink.datastream import StreamExecutionEnvironment\n",
+    "from pyflink.table import StreamTableEnvironment\n",
+    "from pyflink.table.expressions import lit\n",
+    "from pynessie import init\n",
+    "\n",
+    "warehouse = os.path.join(os.getcwd(), \"flink-warehouse\") # where we will store our data\n",
+    "iceberg_flink_runtime_jar = os.path.join(os.getcwd(), \"iceberg-flink-runtime-0.12.0.jar\") # this was downloaded when Binder started, its available on maven central\n",
+    "\n",
+    "env = StreamExecutionEnvironment.get_execution_environment()\n",
+    "env.add_jars(\"file://{}\".format(iceberg_flink_runtime_jar))\n",
+    "table_env = StreamTableEnvironment.create(env)\n",
+    "\n",
+    "nessie_client = init()\n",
+    "\n",
+    "def create_ref_catalog(ref):\n",
+    "    \"\"\"\n",
+    "    Create a flink catalog that is tied to a specific ref.\n",
+    "\n",
+    "    In order to create the catalog we have to first create the branch\n",
+    "    \"\"\"\n",
+    "    hash_ = nessie_client.get_reference(nessie_client.get_default_branch()).hash_\n",
+    "    try:\n",
+    "        nessie_client.create_branch(ref, hash_)\n",
+    "    except:\n",
+    "        pass # already created\n",
+    "    # The important args below are:\n",
+    "    # type: tell Flink to use Iceberg as the catalog\n",
+    "    # catalog-impl: which Iceberg catalog to use, in this case we want Nessie\n",
+    "    # uri: the location of the nessie server.\n",
+    "    # ref: the Nessie ref/branch we want to use (defaults to main)\n",
+    "    # warehouse: the location this catalog should store its data\n",
+    "    table_env.execute_sql(\n",
+    "            f\"\"\"CREATE CATALOG {ref}_catalog WITH (\n",
+    "            'type'='iceberg',\n",
+    "            'catalog-impl'='org.apache.iceberg.nessie.NessieCatalog',\n",
+    "            'uri'='http://localhost:19120/api/v1',\n",
+    "            'ref'='{ref}',\n",
+    "            'warehouse' = '{warehouse}')\"\"\"\n",
+    "        )\n",
+    "create_ref_catalog(nessie_client.get_default_branch())\n",
+    "print(\"Flink running\")"
```

Comment:
This step fails:

```
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/srv/conda/envs/flink-demo/lib/python3.7/site-packages/pyflink/lib/slf4j-log4j12-1.7.25.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/srv/conda/envs/flink-demo/lib/python3.7/site-packages/pyflink/lib/log4j-slf4j-impl-2.12.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
---------------------------------------------------------------------------
Py4JJavaError                             Traceback (most recent call last)
/tmp/ipykernel_74/1981718022.py in <module>
     39             'warehouse' = '{warehouse}')"""
     40         )
---> 41 create_ref_catalog(nessie_client.get_default_branch())
     42 print("Flink running")

/tmp/ipykernel_74/1981718022.py in create_ref_catalog(ref)
     37             'uri'='http://localhost:19120/api/v1',
     38             'ref'='{ref}',
---> 39             'warehouse' = '{warehouse}')"""
     40         )
     41 create_ref_catalog(nessie_client.get_default_branch())

/srv/conda/envs/flink-demo/lib/python3.7/site-packages/pyflink/table/table_environment.py in execute_sql(self, stmt)
    764         """
    765         self._before_execute()
--> 766         return TableResult(self._j_tenv.executeSql(stmt))
    767 
    768     def create_statement_set(self) -> StatementSet:

/srv/conda/envs/flink-demo/lib/python3.7/site-packages/py4j/java_gateway.py in __call__(self, *args)
   1284         answer = self.gateway_client.send_command(command)
   1285         return_value = get_return_value(
-> 1286             answer, self.gateway_client, self.target_id, self.name)
   1287 
   1288         for temp_arg in temp_args:

/srv/conda/envs/flink-demo/lib/python3.7/site-packages/pyflink/util/exceptions.py in deco(*a, **kw)
    145     def deco(*a, **kw):
    146         try:
--> 147             return f(*a, **kw)
    148         except Py4JJavaError as e:
    149             from pyflink.java_gateway import get_gateway

/srv/conda/envs/flink-demo/lib/python3.7/site-packages/py4j/protocol.py in get_return_value(answer, gateway_client, target_id, name)
    326                 raise Py4JJavaError(
    327                     "An error occurred while calling {0}{1}{2}.\n".
--> 328                     format(target_id, ".", name), value)
    329             else:
    330                 raise Py4JError(

Py4JJavaError: An error occurred while calling o11.executeSql.
: org.apache.flink.table.api.NoMatchingTableFactoryException: Could not find a suitable table factory for 'org.apache.flink.table.factories.CatalogFactory' in
the classpath.

Reason: Required context properties mismatch.

The following properties are requested:
catalog-impl=org.apache.iceberg.nessie.NessieCatalog
ref=main
type=iceberg
uri=http://localhost:19120/api/v1
warehouse=/home/jovyan/notebooks/flink-warehouse

The following factories have been considered:
org.apache.flink.table.catalog.GenericInMemoryCatalogFactory
	at org.apache.flink.table.factories.TableFactoryService.filterByContext(TableFactoryService.java:301)
	at org.apache.flink.table.factories.TableFactoryService.filter(TableFactoryService.java:179)
	at org.apache.flink.table.factories.TableFactoryService.findSingleInternal(TableFactoryService.java:140)
	at org.apache.flink.table.factories.TableFactoryService.find(TableFactoryService.java:109)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.createCatalog(TableEnvironmentImpl.java:1119)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeOperation(TableEnvironmentImpl.java:1019)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:666)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.flink.api.python.shaded.py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at org.apache.flink.api.python.shaded.py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at org.apache.flink.api.python.shaded.py4j.Gateway.invoke(Gateway.java:282)
	at org.apache.flink.api.python.shaded.py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at org.apache.flink.api.python.shaded.py4j.commands.CallCommand.execute(CallCommand.java:79)
	at org.apache.flink.api.python.shaded.py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

Reply:
how were you running that? Looks like the flink jar wasnt fetched first

Reply:
Just in mybinder.org

Reply:
ah weird. mybinder seems to have a different launch point than my local repo2docker. Fixed